src/cpp/server/auth/ServerAuthCommon.cpp

```diff
@@ -424,6 +439,60 @@ void prepareHandler(handler::Handler& handler,
    handler.refreshAuthCookies = boost::bind(setSignInCookies, _1, _2, _3, _4);
 }
 
+std::string userIdentifierToLocalUsername(const std::string& userIdentifier)
+{
+   static core::thread::ThreadsafeMap<std::string, std::string> cache;
+   std::string username = userIdentifier;
+
+   if (cache.contains(userIdentifier))
+   {
+      username = cache.get(userIdentifier);
+   }
+   else
+   {
+      // The username returned from this function is eventually used to create
+      // a local stream path, so it's important that it agree with the system
+      // view of the username (as that's what the session uses to form the
+      // stream path), which is why we do a username => username transform
+      // here. See case 5413 for details.
+      core::system::User user;
+      // This call getpwnam - that will return the passwd line that matches this user identifier
+      Error error = core::system::User::getUserFromIdentifier(userIdentifier, user);
+      if (error)
+      {
+         // log the error and return the original user identifier as a fallback
+         error.addProperty("description", "Error converting userIdentifier to username");
+         LOG_ERROR(error);
+      }
+      else
+      {
+         // This gets the passwd entry for the user-id - this is what the session will do so
+         // when a uid is aliased, we need to go back to the uid for the real username
+         error = core::system::User::getUserFromIdentifier(user.getUserId(), user);
```

Comment:
Once you have the user object, haven't you already read the passwd entry? Unless I'm mistaken both `User::getUserFromIdentifier` calls should return a fully populated user object

Reply:
In my testing, if you have an alias where there are two pwd entries for the same uid, getUserFromIdentifier returns the same username in the pw entry - i.e. the alias. So the second call makes sure that we get the same pw entry that the session will get.